### PR TITLE
Add support for ECDSA TSS with safe primes

### PR DESCRIPTION
**Pull request by pranavjain97**

### CODE REVIEW

**Review by sachushaji — COMMENTED** [review]
Build seems to be failing

File: modules/bitgo/test/ecdh.ts [hunk 1]
```diff
@@ -34,4 +38,19 @@ describe('ECDH sharing secret', () => {
       });
     }
   });
+
+  it('signMessageWithDerivedEcdhKey/verifyEcdhSignature', function () {
```
Comment (zahin-mohammad) [inline modules/bitgo/test/ecdh.ts:42]: Please make this title more descriptive

File: modules/bitgo/test/v2/unit/internal/tssUtils/ecdsa.ts [hunk 2]
```diff
@@ -8,29 +8,31 @@ import * as sinon from 'sinon';
...
 import { nockSendSignatureShareWithResponse } from './common';
-import { createWalletSignatures, nockGetChallenge, nockGetTxRequest } from '../../tss/helpers';
-import { ecc } from '@bitgo/utxo-lib';
-const createKeccakHash = require('keccak');
+import { createWalletSignatures, nockGetChallenge, nockGetChallenges, nockGetTxRequest } from '../../tss/helpers';
+import { bip32, ecc } from '@bitgo/utxo-lib';
 import { Hash } from 'crypto';
+import { mockChallengeA, mockChallengeB, mockChallengeC } from './mocks/ecdsaNtilde';
+
+const createKeccakHash = require('keccak');
```
Comment (zahin-mohammad) [inline modules/bitgo/test/v2/unit/internal/tssUtils/ecdsa.ts:35]: can use an import here instead

File: modules/bitgo/test/v2/unit/internal/tssUtils/ecdsa.ts [hunk 3]
```diff
@@ -577,14 +582,21 @@ describe('TSS Ecdsa Utils:', async function () {
...
       ]);
 
-      const bitgoChallenge = await rangeProof.generateNTilde(3072);
+      const serializedEntChallenge = mockChallengeA;
+      const serializedBitgoChallenge = mockChallengeB;
+
+      const deserializedEntChallenge = ecdsaStaticUtils.deserializeNTilde(serializedEntChallenge);
+      const deserializedBitGoChallenge = ecdsaStaticUtils.deserializeNTilde(serializedBitgoChallenge);
+
+      sinon.stub(rangeProof, 'generateNTilde').resolves(deserializedEntChallenge);
```
Comment (zahin-mohammad) [inline modules/bitgo/test/v2/unit/internal/tssUtils/ecdsa.ts:591]:
```suggestion
      sandbox.stub(rangeProof, 'generateNTilde').resolves(deserializedEntChallenge);
```

File: modules/bitgo/test/v2/unit/internal/tssUtils/ecdsa.ts [hunk 4]
```diff
@@ -804,6 +816,300 @@ describe('TSS Ecdsa Utils:', async function () {
...
+        userId: 'id',
+        userEmail: 'user@bitgo.com',
+        derivedPubkey: bip32.fromBase58(adminEcdhKey.xpub).derivePath(derivationPath).publicKey.toString('hex'),
+        derivationPath: derivationPath,
+        ecdhKeychain: 'my keychain',
+      });
+
+      sinon.stub(bitgo, 'getECDHKeychain').resolves({
+        encryptedXprv: encryptedXprv,
+      });
```
Comment (zahin-mohammad) [inline modules/bitgo/test/v2/unit/internal/tssUtils/ecdsa.ts:943]: When you have mutliple tests like this, dont use the global sinon, use the sandbox instead.

File: modules/sdk-api/src/bitgoAPI.ts [hunk 5]
```diff
@@ -1208,6 +1209,20 @@ export class BitGoAPI implements BitGoBase {
     return this.post(this.url('/user/sharingkey')).send({ email }).result();
   }
 
+  /**
+   * Users that want to sign with a key will use this api to fetch the keychain and the path.
+   * Users that want to verify a signature will use this api to fetch another users ecdh pubkey.
+   * @param bitgo
+   * @param enterpriseId
+   * @param userId
```
Comment (zahin-mohammad) [inline modules/sdk-api/src/bitgoAPI.ts:1217]: add a comment her describing what happens if userId is not provided

File: modules/sdk-api/src/bitgoAPI.ts [hunk 6]
```diff
@@ -1524,12 +1539,15 @@ export class BitGoAPI implements BitGoBase {
   /**
    * Gets the user's private keychain, used for receiving shares
    */
-  async getECDHSharingKeychain(): Promise<any> {
-    const result = await this.get(this.url('/user/settings')).result();
-    if (!result.settings.ecdhKeychain) {
-      return new Error('ecdh keychain not found for user');
+  async getECDHKeychain(ecdhKeychainPub?: string): Promise<any> {
```
Comment (zahin-mohammad) [inline modules/sdk-api/src/bitgoAPI.ts:1542]: Update the comment for this fn. It is no longer only for receiving shares. it is also for signing now.

File: modules/sdk-core/src/bitgo/bitgoBase.ts [hunk 7]
```diff
@@ -12,9 +13,10 @@ export interface BitGoBase {
   readonly env: EnvironmentName;
   fetchConstants(): Promise<any>;
   get(url: string): BitGoRequest;
-  getECDHSharingKeychain(): Promise<any>;
+  getECDHKeychain(ecdhKeychainPub?: string): Promise<any>;
```
Comment (zahin-mohammad) [inline modules/sdk-core/src/bitgo/bitgoBase.ts:16]: can we have a better type then any here

File: modules/sdk-core/src/bitgo/enterprise/iEnterprise.ts [hunk 8]
```diff
@@ -12,4 +15,11 @@ export interface IEnterprise {
   getFirstPendingTransaction(params?: Record<string, never>): Promise<any>;
   settlements(): ISettlements;
   affirmations(): IAffirmations;
+  getAndVerifyEcdsaBitGoChallengeProofs(userPassword: string): Promise<BitGoProofSignatures>;
+  uploadAndEnableEcdsaSigning(
```
Comment (zahin-mohammad) [inline modules/sdk-core/src/bitgo/enterprise/iEnterprise.ts:19]: enabling signing is a side effect of this fn, but what it actually does is just upload their challenge.

```suggestion
  uploadTssEcdsaChallenge(
```

File: modules/sdk-core/src/bitgo/utils/tss/ecdsa/ecdsa.ts [hunk 9]
```diff
@@ -771,6 +784,67 @@ export class EcdsaUtils extends baseTSSUtils<KeyShare> {
...
+   * @param walletId
+   * @param enterpriseId
+   */
+  static async getChallengesForEcdsaSigning(
+    bitgo: BitGoBase,
+    walletId: string,
+    enterpriseId: string
+  ): Promise<ApiChallenges> {
+    const urlPath = `/wallet/${walletId}/challenges`;
+    const result = await bitgo.get(bitgo.url(urlPath, 2)).query({}).result();
```
Comment (zahin-mohammad) [inline modules/sdk-core/src/bitgo/utils/tss/ecdsa/ecdsa.ts:800]: can this be delegated to a function on the wallet class? It's a little odd to do it here.

File: modules/sdk-core/src/bitgo/ecdh.ts [hunk 10]
```diff
@@ -63,3 +63,21 @@ export function getSharedSecret(
   assert.strictEqual(buffer.length, 32);
   return buffer;
 }
+
+/**
+ Signs a message using a given ecdh xprv at a given path
+ */
+export function signMessageWithDerivedEcdhKey(message: string, xprv: string, path: string): Buffer {
```
Comment (zahin-mohammad) [inline modules/sdk-core/src/bitgo/ecdh.ts:70]: Let's add unit tests for these new functions.

**Review by zahin-mohammad — CHANGES_REQUESTED** [review]

File: modules/bitgo/test/v2/unit/internal/tssUtils/ecdsa.ts (hunk 2 above)
Comment (pranavjain97) [reply modules/bitgo/test/v2/unit/internal/tssUtils/ecdsa.ts:35]: unrelated change

**Review by rlutzbitgo — APPROVED** [review]
Thought Zahin's comments were good, also agree that throwing on attempting to retrieve an eddsa challenge makes sense at least until we switch primes. Not blocking over it though. Nice job!

**Review by zahin-mohammad — APPROVED** [review]

**Comment by pranavjain97** [comment]
>agree that throwing on attempting to retrieve an eddsa challenge makes sense at least until we switch primes

We already have a check for ecdsa curve before retriveing a challenge on the `/challenge` api it uses, so we should be fine there.

**Pull request merged**